wrappers/pkcs11/pkcs11_client.go

```diff
@@ -0,0 +1,632 @@
+	if err != nil {
+		return 0, fmt.Errorf("failed to open session: %w", err)
+	}
+	err = c.client.Login(session, pkcs11.CKU_USER, c.pin)
+	if err != nil {
+		return 0, fmt.Errorf("failed to login: %w", err)
+	}
+	return session, nil
+}
+
+func (c *Pkcs11Client) CloseSession(session pkcs11.SessionHandle) {
+	if c.client == nil {
+		return
+	}
+	c.client.Logout(session)
+	c.client.CloseSession(session)
+}
+
+// Find on key for the given Label, ID and Mechanism.
+func (c *Pkcs11Client) FindKey(session pkcs11.SessionHandle, key Pkcs11Key, typ uint) (pkcs11.ObjectHandle, error) {
+	template := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_LABEL, []byte(key.label)),
+		pkcs11.NewAttribute(typ, true),
+	}
+	if keyIdBytes, err := hex.DecodeString(key.id); err == nil {
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_ID, keyIdBytes))
+	}
+	if c.mechanism != 0 {
+		keyTypeString, err := GetKeyTypeFromMech(c.mechanism)
+		if err != nil {
+			return 0, fmt.Errorf("failed to get key type from mechanism: %s", err)
+		}
+		template = append(template, pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, keyTypeString))
+	}
+
+	if err := c.client.FindObjectsInit(session, template); err != nil {
+		return 0, fmt.Errorf("failed to pkcs11 FindObjectsInit: %s", err)
+	}
+	obj, _, err := c.client.FindObjects(session, 2)
+	if err != nil {
+		return 0, fmt.Errorf("failed to pkcs11 FindObjects: %s", err)
+	}
+	if err := c.client.FindObjectsFinal(session); err != nil {
+		return 0, fmt.Errorf("failed to pkcs11 FindObjectsFinal: %s", err)
+	}
+	if len(obj) == 0 {
+		return 0, fmt.Errorf("no key found for the label: %s", key.label)
+	}
+	if len(obj) != 1 {
+		return 0, fmt.Errorf("got more than 1 key for the label: %s", key.label)
+	}
+
+	return obj[0], nil
+}
+
+func (c *Pkcs11Client) GetKeyMechanism(session pkcs11.SessionHandle, key pkcs11.ObjectHandle) (uint, error) {
+	template := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_KEY_TYPE, nil),
+	}
+	attr, err := c.client.GetAttributeValue(session, pkcs11.ObjectHandle(key), template)
+	if err != nil {
+		return 0, fmt.Errorf("failed to pkcs11 GetAttributeValue: %s", err)
+	}
+
+	attrMap := GetAttributesMap(attr)
+	keyType := GetValueAsInt(attrMap[pkcs11.CKA_KEY_TYPE])
+
+	mechanism := uint(0)
+	switch keyType {
+	case pkcs11.CKK_AES:
+		if c.mechanism != 0 {
+			mechanism = c.mechanism
+		} else {
+			mechanism = pkcs11.CKM_AES_GCM
+		}
+	case pkcs11.CKK_RSA:
+		if c.mechanism != 0 {
+			mechanism = c.mechanism
+		} else {
+			mechanism = pkcs11.CKM_RSA_PKCS_OAEP
+		}
+	default:
+		return 0, fmt.Errorf("unsupported key type: %d", keyType)
+	}
+
+	return mechanism, nil
+}
+
+func (c *Pkcs11Client) GetMechAndPrefixSizeForKey(session pkcs11.SessionHandle, key pkcs11.ObjectHandle) (uint, int, error) {
+	mechanism, err := c.GetKeyMechanism(session, key)
+	if err != nil {
+		return 0, 0, err
+	}
+	switch mechanism {
+	case pkcs11.CKM_AES_GCM:
+		return mechanism, 16, nil
```

Comment:
@glatigny Now _that_ is unintuitive. The panic occurs in SoftHSM because we passed 16 bytes instead of 12 here.